### PR TITLE
Improve rand tests

### PR DESCRIPTION
**Pull request by vprusso**
Started off doing some improvements for the testing in `rand/` and hit a few additional areas. Namely:

- Adding parameterization to a subset of tests
- Reformatting lines based on the new longer linting standard
- Fixing both `inner_product` and `outer_product` functions (these most likely can be removed as they are just numpy wrappers)
- Adding more comprehensive docstrings to `vectors_from_gram_matrix` and `gram_matrix_to_vectors` functions.
- Minor simplifications and cleanup. 
- Increasing coverage of tensor product function.

### CODE REVIEW

**Comment by vprusso** [comment]
Sorry for all the noise on this one. I'll be leaving this one alone for now so it stays stable. Should be good to be reviewed, @purva-thakre !

**Comment by vprusso** [comment]
Just a ping for this PR and https://github.com/vprusso/toqito/pull/271 to see if we want to get these in before some of the other PRs coming down the pipe (might make the merging easier if we chuck these in first) fyi @purva-thakre

File: toqito/states/werner.py [hunk 1]
```diff
@@ -94,14 +87,9 @@ def werner(dim: int, alpha: float | list[float]) -> np.ndarray:
...
     """
-    # The total number of permutation operators.
-    if isinstance(alpha, float):
-        n_fac = 2
-    else:
-        n_fac = len(alpha) + 1
-
     # Multipartite Werner state.
-    if n_fac > 2:
+    if isinstance(alpha, list):
```
Comment (purva-thakre) [inline toqito/states/werner.py:91]: Flagged by pylint. 

https://github.com/vprusso/toqito/actions/runs/6922210428/job/18828672060#step:6:13

File: toqito/matrix_ops/tests/test_inner_product.py [hunk 2]
```diff
@@ -1,44 +1,28 @@
...
+
+@pytest.mark.parametrize("v1, v2", [
+    # Different dimensions of vectors.
+    (np.array([1, 2, 3]), np.array([4, 5, 6, 7])),
+    # Vector and 2D array.
+    (np.array([1, 2, 3]), np.array([[4, 5, 6], [7, 8, 9]])),
+])
+def test_inner_product_invalid_input(v1, v2):
     with np.testing.assert_raises(ValueError):
         inner_product(v1, v2)
```
Comment (purva-thakre) [inline toqito/matrix_ops/tests/test_inner_product.py:28]:
```suggestion
    with pytest.raises(ValueError):
        inner_product(v1, v2)
```

File: toqito/matrix_ops/tests/test_outer_product.py [hunk 3]
```diff
@@ -1,36 +1,26 @@
...
-    v1, v2 = np.array([1, 2, 3]), np.array([[4, 5, 6], [7, 8, 9]])
+@pytest.mark.parametrize("v1, v2", [
+    # Different dimensions of vectors.
+    (np.array([1, 2, 3]), np.array([4, 5, 6, 7])),
+    # Vector and 2D array.
+    (np.array([1, 2, 3]), np.array([[4, 5, 6], [7, 8, 9]])),
+])
+def test_outer_product_invalid_input(v1, v2):
     with np.testing.assert_raises(ValueError):
         outer_product(v1, v2)
```
Comment (purva-thakre) [inline toqito/matrix_ops/tests/test_outer_product.py:26]:
```suggestion
    with pytest.raises(ValueError):
        outer_product(v1, v2)
```

File: toqito/matrix_ops/tests/test_outer_product.py [hunk 4]
```diff
@@ -1,36 +1,26 @@
...
-def test_outer_product():
-    """Test with two vectors, no complications."""
+@pytest.mark.parametrize("v1, v2, expected_result", [
+    # Test with two vectors, no complications.
+    (np.array([1, 2, 3]), np.array([4, 5, 6]), np.array([[4, 5, 6], [8, 10, 12], [12, 15, 18]])),
+    # Test with two vectors, with negative input/output values.
+    (np.array([-1, 2, 3]), np.array([4, 5, 6]), np.array([[-4, -5, -6], [8, 10, 12], [12, 15, 18]])),
+])
+def test_outer_product(v1, v2, expected_result):
+    np.testing.assert_equal(outer_product(v1, v2), expected_result)
```
Comment (purva-thakre) [inline toqito/matrix_ops/tests/test_outer_product.py:15]:
```suggestion
    np.testing.assert_array_equal(outer_product(v1, v2), expected_result)
```

File: toqito/matrix_ops/tests/test_tensor.py [hunk 5]
```diff
@@ -150,7 +150,28 @@ def test_tensor_list_3():
...
+    # Call the tensor function
+    result = tensor(matrices)
+
+    # Assert that the result is as expected
+    np.testing.assert_array_equal(result, expected_output)
+
+
 def test_tensor_empty_args():
     r"""Test tensor with no arguments."""
     with np.testing.assert_raises(ValueError):
```
Comment (purva-thakre) [inline toqito/matrix_ops/tests/test_tensor.py:176]:
```suggestion
    with pytest.raises(ValueError):
```

File: toqito/matrix_ops/tests/test_vectors_from_gram_matrix.py [hunk 6]
```diff
@@ -24,3 +32,14 @@ def test_vectors_from_gram_matrix_not_psd():
...
+
+
+@pytest.mark.parametrize("gram", [
+    # Non-square matrix.
+    (np.array([[1, 2], [4, 5], [7, 8]])),
+    # Non-symmetric matrix.
+    (np.array([[1, 2], [3, 4]])),
+])
+def test_vectors_from_gram_matrix_invalid_input(gram):
+    with np.testing.assert_raises(np.linalg.LinAlgError):
```
Comment (purva-thakre) [inline toqito/matrix_ops/tests/test_vectors_from_gram_matrix.py:44]:
```suggestion
    with pytest.raises(np.linalg.LinAlgError):
```

File: toqito/matrix_ops/tests/test_vectors_to_gram_matrix.py [hunk 7]
```diff
@@ -1,18 +1,32 @@
...
+    vectors = vectors_to_gram_matrix(trine)
+    np.testing.assert_allclose(vectors, expected_result)
+
+
+@pytest.mark.parametrize("vectors", [
+    # Vectors of different sizes. 
+    ([np.array([1, 2, 3]), np.array([1, 2])]),
+])
+def test_vectors_to_gram_matrix_invalid_input(vectors):
+    with np.testing.assert_raises(ValueError):
```
Comment (purva-thakre) [inline toqito/matrix_ops/tests/test_vectors_to_gram_matrix.py:31]:
```suggestion
    with pytest.raises(ValueError):
```

File: toqito/rand/tests/test_random_density_matrix.py [hunk 8]
```diff
@@ -1,35 +1,54 @@
...
+    if k_param == dim:
+        mat = random_density_matrix(
+            dim=dim, is_real=is_real, k_param=k_param, distance_metric=distance_metric
+        )
+        np.testing.assert_equal(is_density(mat), True)
+
+
+def is_positive_semidefinite(matrix):
+    """Check if a matrix is positive semidefinite."""
+    return np.all(np.linalg.eigvals(matrix) >= 0)
```
Comment (purva-thakre) [inline toqito/rand/tests/test_random_density_matrix.py:29]: I might be mistaken but where is `matrix` defined?

File: toqito/rand/tests/test_random_unitary.py [hunk 9]
```diff
@@ -1,23 +1,36 @@
...
-    np.testing.assert_equal(is_unitary(mat), True)
+@pytest.mark.parametrize("dim_n", range(2, 8))
+@pytest.mark.parametrize("dim_m", range(2, 8))
+@pytest.mark.parametrize("is_real", [True, False])
+def test_random_unitary_list_dims(dim_n, dim_m, is_real):
+    if dim_n == dim_m:
+        mat = random_unitary(
+            dim=[dim_n, dim_m], is_real=is_real
+        )
+        np.testing.assert_equal(is_unitary(mat), True)
```
Comment (purva-thakre) [inline toqito/rand/tests/test_random_unitary.py:26]:
```suggestion
        assert is_unitary(mat)
```

File: toqito/rand/tests/test_random_unitary.py [hunk 10]
```diff
@@ -1,23 +1,36 @@
...
-def test_random_unitary_not_real():
-    """Generate random non-real unitary matrix."""
-    mat = random_unitary(2)
+@pytest.mark.parametrize("dim", range(2, 8))
+@pytest.mark.parametrize("is_real", [True, False])
+def test_random_unitary_int_dim(dim, is_real):
+    mat = random_unitary(
+        dim=dim, is_real=is_real
+    )
     np.testing.assert_equal(is_unitary(mat), True)
```
Comment (purva-thakre) [inline toqito/rand/tests/test_random_unitary.py:15]:
```suggestion
  assert is_unitary(mat)
```

File: toqito/rand/tests/test_random_state_vector.py [hunk 11]
```diff
@@ -1,40 +1,24 @@
...
+    # Check that real state vector with k_param > 0.
+    (2, True, 1),
+    # Check that real state vector from random state vector is pure.
+    (2, True, 0),
+])
+def test_random_state_vector(dim, is_real, k_param):
+    # We expect the density matrix of any random state vector to be pure.
+    vec = random_state_vector(dim=dim, is_real=is_real, k_param=k_param).reshape(-1, 1)
+    mat = vec @ vec.conj().T
     np.testing.assert_equal(is_pure(mat), True)
```
Comment (purva-thakre) [inline toqito/rand/tests/test_random_state_vector.py:24]:
```suggestion
    assert is_pure(mat)
```

File: toqito/states/tests/test_werner.py [hunk 12]
```diff
@@ -17,7 +19,25 @@ def test_werner_multipartite():
...
+    (2, [0.5, 0.6, 0.7]),
+    # Test with an integer (which is not a valid type for alpha)
+    (2, 5),
+    # Test with a string (which is not a valid type for alpha)
+    (2, "invalid"),
+    # Test with a dictionary (which is not a valid type for alpha)
+    (2, {"key": "value"}),
+])
+def test_werner_state_invalid(dim, alpha):
     with np.testing.assert_raises(ValueError):
```
Comment (purva-thakre) [inline toqito/states/tests/test_werner.py:42]:
```suggestion
    with pytest.raises(ValueError):
```

**Review by purva-thakre — APPROVED** [review]
LGTM except for a few minor changes. 

I think we do want to use `pytest.raises` to test if an error is raised correctly. One thing I noticed after working on some of the issues in #217 is, if numpy is used to check things like this, the covered line might show up as uncovered to pytest.

File: toqito/rand/tests/test_random_density_matrix.py (hunk 8 above)
Comment (vprusso) [reply toqito/rand/tests/test_random_density_matrix.py:29]: Not mistaken. Some weird copy/paste thing here on my part. Removing, nice catch!

File: toqito/states/werner.py (hunk 1 above)
Comment (vprusso) [reply toqito/states/werner.py:91]: Good call! Fixed.

**Comment by vprusso** [comment]
Thanks again for the review, @purva-thakre !

**Pull request merged**